pfcpiface/conn.go

```diff
@@ -82,52 +93,65 @@ func pfcpifaceMainLoop(upf *upf, accessIP, coreIP, sourceIP, smfName string) {
 		go pconn.manageSmfConnection(sourceIP, accessIP, smfName, conn, cpConnectionStatus)
 	}
 
-	// Initialize pkt buf
-	buf := make([]byte, PktBufSz)
+	pfcpRcvdPktsChan := make(chan *rcvdPacket, 1000)
+	pfcpParsedPktsChan := make(chan *parsedPacket, 1000)
 	// Initialize pkt header
 
-	for {
-		err := conn.SetReadDeadline(time.Now().Add(readTimeout))
-		if err != nil {
-			log.Fatalln("Unable to set deadline for read:", err)
-		}
-		// blocking read
-		n, addr, err := conn.ReadFrom(buf)
-		if err != nil {
-			if err, ok := err.(net.Error); ok && err.Timeout() {
-				// do nothing for the time being
-				log.Println(err)
-				cpConnected = false
-				if manageConnection {
-					cpConnectionStatus <- cpConnected
+	pfcpPacketReader := func(pfcpRcvdPkts chan *rcvdPacket) {
+		for {
+			err := conn.SetReadDeadline(time.Now().Add(readTimeout))
+			if err != nil {
+				log.Fatalln("Unable to set deadline for read:", err)
+			}
+			pkt := new(rcvdPacket)
+			// blocking read
+			n, addr, err := conn.ReadFrom(pkt.Buf[:1500])
+			if err != nil {
+				if err, ok := err.(net.Error); ok && err.Timeout() {
+					// do nothing for the time being
+					log.Println(err)
+					cpConnected = false
+					if manageConnection {
+						cpConnectionStatus <- cpConnected
+					}
+					cleanupSessions()
+					continue
 				}
-				cleanupSessions()
-				continue
+				log.Fatalln("Read error:", err)
 			}
-			log.Fatalln("Read error:", err)
+			pkt.Pkt_size = n
+			pkt.Address = addr
+			pfcpRcvdPktsChan <- pkt
 		}
+	}
 
+	pfcpPacketParsing := func(pkt *rcvdPacket, pfcpParsedPktsChan chan *parsedPacket) {
```

Comment:
can make a full fledged function. goreportcard has been complaining about pfcpmainloop for a while now
https://goreportcard.com/report/github.com/omec-project/upf-epc#gocyclo